### ISSUE TITLE
Calling AsyncCommand.ExecuteAsync from unit test (NUnit 3)

### ISSUE THREAD

**Issue by kevkeller** [StephenCleary/Mvvm.Async#22]
Maybe I've missed something, but I can't get AsyncCommands to work with my tests. As soon as ExecuteAsync is called from a test, an InvalidOperationException is thrown.

```System.InvalidOperationException : The calling thread cannot access this object because a different thread owns it.
   at Nito.Mvvm.ThreadAffinity.VerifyCurrentThread()
   at Nito.Mvvm.StrongCanExecuteChanged.OnCanExecuteChanged()
   at Nito.Mvvm.AsyncCommand.<ExecuteAsync>d__11.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at AsyncCommandNUnit.TestClass.<Throws>d__1.MoveNext() in C:\Projects\AsyncCommandNUnit\AsyncCommandNUnit\TestClass.cs:line 24
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at NUnit.Framework.Internal.AsyncInvocationRegion.AsyncTaskInvocationRegion.WaitForPendingOperationsToComplete(Object invocationResult) in C:\src\nunit\nunit\src\NUnitFramework\framework\Internal\AsyncInvocationRegion.cs:line 113
   at NUnit.Framework.Internal.Commands.TestMethodCommand.RunAsyncTestMethod(TestExecutionContext context) in C:\src\nunit\nunit\src\NUnitFramework\framework\Internal\Commands\TestMethodCommand.cs:line 96
```

Here's me test code:
```csharp
    [TestFixture]
    public class TestClass
    {
        [Test]
        public async Task Throws()
        {
            AsyncCommand command = new AsyncCommand(
                async () => { await Task.Delay(1000); });
            await command.ExecuteAsync(null);
        }

        [Test]
        public async Task DoesNotThrow()
        {
            AsyncCommand command = new AsyncCommand(
                () =>
                {
                    Thread.Sleep(1000);
                    return Task.FromResult(0);
                });

            await command.ExecuteAsync(null);
        }

        [Test]
        public async Task AlsoThrows()
        {
            AsyncCommand command = new AsyncCommand(
                () => Task.Delay(1000));

            await command.ExecuteAsync(null);
        }
    }
```

I've already tried explicitly setting Synchronization context without success. Also I'm experiencing the same behavior when executing similar code in a simple console app.

Here's my environment:
```xml
<packages>
  <package id="Nito.Disposables" version="1.2.3" targetFramework="net471" />
  <package id="Nito.Mvvm.Async" version="1.0.0-pre-03" targetFramework="net471" />
  <package id="Nito.Mvvm.Core" version="1.2.2" targetFramework="net471" />
  <package id="NUnit" version="3.10.1" targetFramework="net471" />
  <package id="System.Collections.Immutable" version="1.4.0" targetFramework="net471" />
</packages>
```
running on Win7

**Comment by StephenCleary** [comment]
`Nito.Mvvm` assumes it is running on a single-threaded context, like a UI context. To put it another way, all of the types have a "thread affinity" - they can only be accessed from the thread they were created on. The exception you're seeing is a cross-thread exception.

Using `async`/`await` in both NUnit and Console apps can switch threads. In order to prevent this, you'll need to use an `async`-compatible single-threaded context like [`AsyncContext`](https://www.nuget.org/packages/Nito.AsyncEx.Context/). Something like this:

```c#
[Test]
public void Fixed()
{
  AsyncContext.Run(async () =>
  {
    AsyncCommand command = new AsyncCommand(
        async () => { await Task.Delay(1000); });
    await command.ExecuteAsync(null);
  });
}

**Comment by kevkeller** [comment]
Hm, I thought not using ```ConfigureAwait(false)``` would prevent switching threads. I've already been playing around with ```[SingleThreaded]``` and ```[Apartment(ApartmentState.STA)]``` to no avail, but your solution seems to do the trick.

Thanks a lot!

**Comment by mlimber** [comment]
🧟‍♀️ Zombie thread resurrected for posterity. 🧟‍

Another option is to remove thread affinity requirement by duplicating Nito.Mvvm.Core's `StringCanExecuteChanged` (and/or `WeakCanExecuteChanged`) to remove thread affinity:
```cs
public sealed class AsyncCommandCanExecuteChanged : ICanExecuteChanged
{
    /// <summary>
    /// A factory function for creating CanExecuteChanged objects.
    /// </summary>
    /// <param name="sender">The object to emit the events.</param>
    /// <returns>The newly created <see cref="AsyncCommandCanExecuteChanged"/> object.</returns>
    public static ICanExecuteChanged Create( object sender )
    {
        return new AsyncCommandCanExecuteChanged( sender );
    }
        
    /// <summary>
    /// The sender of the <c>ICommand.CanExecuteChanged</c> event.
    /// </summary>
    private readonly object _sender;

    /// <summary>
    /// The collection of delegates for <see cref="CanExecuteChanged"/>.
    /// </summary>
    private event EventHandler CanExecuteChangedEvent;

    /// <summary>
    /// Creates a new strong-event implementation of <c>ICommand.CanExecuteChanged</c>.
    /// </summary>
    /// <param name="sender">The sender of the <c>ICommand.CanExecuteChanged</c> event.</param>
    public AsyncCommandCanExecuteChanged( object sender )
    {
        _sender = sender;
    }

    /// <inheritdoc />
    /// <summary>
    /// Provides notification that the result of <c>ICommand.CanExecute</c> may be different.
    /// </summary>
    public event EventHandler CanExecuteChanged
    {
        add => CanExecuteChangedEvent += value;
        remove => CanExecuteChangedEvent -= value;
    }

    /// <inheritdoc />
    /// <summary>
    /// Raises the <see cref="CanExecuteChanged" /> event.
    /// </summary>
    public void OnCanExecuteChanged()
    {
        CanExecuteChangedEvent?.Invoke( _sender, EventArgs.Empty );
    }
}
```
And then pass it into your commands at construction, e.g.:
```cs
DeleteSelectedObjectsCommand = new CustomAsyncCommand( 
    DeleteSelectedObjectsAsync,
    _ => HasSelection,
    AsyncCommandCanExecuteChanged.Create );
```
Then the test can await as usual:
```cs
[Test]
public async Task TestDeletion()
{
    // ... assembly here ...
    await _viewModel.DeleteSelectedObjectsCommand.ExecuteAsync( null );
    // ... asserts here ...
}
```
Of course you may find you get bit by thread affinity being off, but it could also be not required in your context, as (I think) it is in mine. (Perhaps Stephen will reply that this is ill-advised, but I'll wait and see. :-) )

**Comment by StephenCleary** [comment]
Well, yeah, the thread affinity was deliberately put in because lots of the code assumes it is running on the same thread. More specifically, events are raised using the simple pattern rather than using an extra variable, and all private state variables are completely unprotected (no `lock`, `volatile`, or barriers of any kind).

So, disabling the thread affinity guard voids your warranty. :)

I would still recommend using `AsyncContext.Run`, or an alternative test framework like xUnit that provides its own context.

**Comment by mlimber** [comment]
Ah, so you think this is a defect in NUnit?

**Comment by StephenCleary** [comment]
Not a defect per se. Just a different way of doing things that makes testing asynchronous UI code harder.

**Comment by mlimber** [comment]
It appears the NUnit team agrees with you. :-) They've [fixed it in 3.11](https://github.com/nunit/nunit/issues/1200).

**Comment by StephenCleary** [comment]
Excellent! I remember recommending that when v3 was in beta, but they didn't want to make that change at the time.
